src/Microsoft.AspNet.Mvc.Razor/Precompilation/CodeAnalysisSymbolBasedTypeInfo.cs

```diff
@@ -0,0 +1,251 @@
+            {
+                dictionaryInterface = (INamedTypeSymbol)_type;
+            }
+            else
+            {
+                dictionaryInterface = _type
+                    .AllInterfaces
+                    .FirstOrDefault(implementedInterface => implementedInterface.ConstructedFrom == dictionarySymbol);
+            }
+
+            if (dictionaryInterface != null)
+            {
+                Debug.Assert(dictionaryInterface.TypeArguments.Length == 2);
+
+                return new[]
+                {
+                    GetFullName(dictionaryInterface.TypeArguments[0]),
+                    GetFullName(dictionaryInterface.TypeArguments[1])
+                };
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Gets the assembly qualified named of the specified <paramref name="symbol"/>.
+        /// </summary>
+        /// <param name="symbol">The <see cref="ITypeSymbol" /> to generate the name for.</param>
+        /// <returns>The assembly qualified name.</returns>
+        public static string GetAssemblyQualifiedName([NotNull] ITypeSymbol symbol)
+        {
+            var builder = new StringBuilder();
+            GetAssemblyQualifiedName(builder, symbol);
+
+            return builder.ToString();
+        }
+
+        private List<IPropertyInfo> PopulateProperties()
+        {
+            var properties = new List<IPropertyInfo>();
+            var overridenProperties = new HashSet<IPropertySymbol>();
+            var type = _type;
+
+            while (type != null)
+            {
+                foreach (var member in type.GetMembers())
+                {
+                    if (member.Kind == SymbolKind.Property)
+                    {
+                        var propertySymbol = (IPropertySymbol)member;
+                        if (!propertySymbol.IsIndexer && !overridenProperties.Contains(propertySymbol))
+                        {
+                            var propertyInfo = new CodeAnalysisSymbolBasedPropertyInfo(propertySymbol, _symbolLookup);
+                            properties.Add(propertyInfo);
+
+                            if (propertySymbol.IsOverride)
+                            {
+                                overridenProperties.Add(propertySymbol.OverriddenProperty);
+                            }
+                        }
+                    }
+                }
+
+                type = type.BaseType;
+            }
+
+            return properties;
+        }
+
+        private static string GetFullName(ITypeSymbol typeSymbol)
+        {
+            var nameBuilder = new StringBuilder();
+            GetFullName(nameBuilder, typeSymbol);
+
+            return nameBuilder.Length == 0 ? null : nameBuilder.ToString();
+        }
+
+        private static void GetFullName(StringBuilder nameBuilder, ITypeSymbol typeSymbol)
+        {
+            if (typeSymbol.Kind == SymbolKind.TypeParameter)
```

Comment:
Yup. `IDictionary<,>` produces the name `IDictionary;2` - this is to accommodate scenarios along those lines.